core/src/test/java/org/apache/calcite/test/TestMetadataHandlers.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexSimplify;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.util.Sarg;
+import org.apache.calcite.util.TimestampString;
+import org.apache.calcite.util.Util;
+
+import com.google.common.collect.Range;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Metadata handlers that are used in {@link MaterializedViewRelOptRulesTest}.
+ */
+public class TestMetadataHandlers {
+  /**
+   * Modify the rowCount of the materialized view to be lower than the base table.
+   */
+  public static class TestRelMdRowCount extends RelMdRowCount {
+    public static final RelMetadataProvider SOURCE =
+        ReflectiveRelMetadataProvider.reflectiveSource(
+            new TestRelMdRowCount(), BuiltInMetadata.RowCount.Handler.class);
+
+    public Double getRowCount(TableScan rel, RelMetadataQuery mq) {
+      if (rel.getTable().getQualifiedName().toString().equalsIgnoreCase("[hr, events]")) {
+        return 100000d;
+      } else {
+        return 100d;
+      }
+    }
+  }
+
+  /**
+   * Modify the selectivity of SArg to be more selective for a smaller range.
+   */
+  @SuppressWarnings("BetaApi")
+  public static class TestRelMdSelectivity extends RelMdSelectivity {
+
+    public static final RelMetadataProvider SOURCE =
+        ReflectiveRelMetadataProvider.reflectiveSource(
+            new TestRelMdSelectivity(), BuiltInMetadata.Selectivity.Handler.class);
+
+    public Double getSelectivity(RelNode rel, RelMetadataQuery mq,
+        @Nullable RexNode predicate) {
+      if (predicate == null) {
+        return RelMdUtil.guessSelectivity(predicate);
+      }
+      final RexExecutor executor =
+          Util.first(rel.getCluster().getPlanner().getExecutor(), RexUtil.EXECUTOR);
+      final RexSimplify rexSimplify = new RexSimplify(rel.getCluster().getRexBuilder(),
+          RelOptPredicateList.EMPTY, executor);
+      RexNode e = rexSimplify.simplify(predicate);
+      if (e.getKind() == SqlKind.SEARCH) {
+        RexCall call = (RexCall) e;
+        Sarg sarg = ((RexLiteral) call.getOperands().get(1)).getValueAs(Sarg.class);
+        double selectivity = 0.d;
+        for (Object o : sarg.rangeSet.asRanges()) {
+          Range r = (Range) o;
+          if (!r.hasLowerBound() || !r.hasUpperBound()
+              || !(r.lowerEndpoint() instanceof TimestampString)
+              || !(r.lowerEndpoint() instanceof TimestampString)) {
+            // ignore predicates where the type isn't a timestamp
+            return RelMdUtil.guessSelectivity(predicate);
+          }
+          long lowerBound = ((TimestampString) r.lowerEndpoint()).getMillisSinceEpoch();
+          long upperBound = ((TimestampString) r.upperEndpoint()).getMillisSinceEpoch();
+          // only used for a range less than one day
+          selectivity += (upperBound - lowerBound) / (Double.valueOf(DateTimeUtils.MILLIS_PER_DAY));
```

Comment:
Unnecessary boxing, it can be simplified as:
```suggestion
          selectivity += (upperBound - lowerBound) / ((double) DateTimeUtils.MILLIS_PER_DAY);
```

Reply:
done